db/memtable.cc

```diff
@@ -532,38 +532,272 @@ Status MemTable::VerifyEncodedEntry(Slice encoded,
-  const uint32_t encoded_len = VarintLength(internal_key_size) +
-                               internal_key_size + VarintLength(val_size) +
-                               val_size;
-  char* buf = nullptr;
-  std::unique_ptr<MemTableRep>& table =
-      type == kTypeRangeDeletion ? range_del_table_ : table_;
-  KeyHandle handle = table->Allocate(encoded_len, &buf);
+  *encoded_len = VarintLength(internal_key_size) + internal_key_size +
+                 VarintLength(val_size) + val_size;
+  KeyHandle handle = table->Allocate(*encoded_len, out_buf);
+  char* buf = *out_buf;
 
   char* p = EncodeVarint32(buf, internal_key_size);
   memcpy(p, key.data(), key_size);
-  Slice key_slice(p, key_size);
+  if (key_slice) {
+    *key_slice = Slice(p, key_size);
+  }
   p += key_size;
   uint64_t packed = PackSequenceAndType(s, type);
   EncodeFixed64(p, packed);
   p += 8;
   p = EncodeVarint32(p, val_size);
-  memcpy(p, value.data(), val_size);
-  assert((unsigned)(p + val_size - buf) == (unsigned)encoded_len);
+  if (tombstone_seqs.empty()) {
+    memcpy(p, value.data(), value.size());
+  } else {
+    p = EncodeVarint32(p, static_cast<uint32_t>(tombstone_seqs.size()));
+    memcpy(p, value.data(), value.size());
+    p += value.size();
+    for (auto& i : tombstone_seqs) {
+      EncodeFixed64(p, i);
+      p += 8;
+    }
+  }
+  return handle;
+}
+
+std::optional<SequenceNumber> MemTable::MaxCoveringTombstoneSeqnum(
+    const ReadOptions& read_options, const LookupKey& key) {
+  if (read_options.ignore_range_deletions ||
+      is_range_del_table_empty_.load(std::memory_order_relaxed)) {
+    return std::nullopt;
+  }
+  std::unique_ptr<MemTableIterator> iter(
+      new MemTableIterator(*this, read_options, nullptr /* arena */,
+                           true /* use_range_del_table */));
+  if (iter == nullptr) {
+    return std::nullopt;
+  }
+
+  Slice user_key = key.user_key();
+  ParsedInternalKey seek_internal_key(user_key, 0, kTypeRangeDeletion);
+  std::string seek_key;
+  AppendInternalKey(&seek_key, seek_internal_key);
+  iter->SeekForPrev(Slice(seek_key.data(), seek_key.size()));
+  if (!iter->Valid()) {
+    return std::nullopt;
+  }
+  Slice tombstone_end_key = iter->value();
+  FragmentedRangeTombstoneList::DecodeRangeTombstoneValue(&tombstone_end_key, nullptr);
+  if (comparator_.comparator.user_comparator()->Compare(user_key, tombstone_end_key) >= 0) {
+    return std::nullopt;
+  }
+
+  ParsedInternalKey k(ExtractUserKey(iter->key()), kMaxSequenceNumber, kTypeRangeDeletion);
+  std::string s;
+  AppendInternalKey(&s, k);
+  iter->Seek(Slice(s.data(), s.size()));
+
+  tombstone_end_key = iter->value();
+  std::vector<SequenceNumber> tombstone_seqs;
+  FragmentedRangeTombstoneList::DecodeRangeTombstoneValue(&tombstone_end_key, &tombstone_seqs);
+  SequenceNumber read_seq = GetInternalKeySeqno(key.internal_key());
+  for (auto t: tombstone_seqs) {
+    if (t <= read_seq) {
+      return t;
+    }
+  }
```

Comment:
Not necessarily in this PR: since sequence numbers are ordered, we can optimize this part by decoding as needed and stop at the first sequence number <= read_seq? I guess we can also do binary search here :)

Reply:
I was thinking the binary search could happen during decoding sequence numbers, so that we only do `log(n)` decoding. But I'm fine with the current approach too and we can make it a potential future TODO.